Taxonomy

simonvh opened the pull request:
* Added the functionality to search by taxonomy_id
* Added the functionality to retrieve taxonomy_id for a genome name
* Added the functionality to retrieve assembly_id for a genome name
* The README.txt file of the genomes will now contains the species taxonomy_id and the assembly accession id. These are changes that later can be used to compare genomes between providers.

siebrenf commented on genomepy/provider.py:
```diff
@@ -804,7 +964,10 @@ class NCBIProvider(ProviderBase):
+        # Necessary for bucketcache, otherwise methods will identical names
+        # from different classes will use the same cache :-O!
+        self.name = "NCBI"
+        self.genomes = self._get_genomes()
```
This is a bit annoying as it starts the moment you begin using genomepy.
Could NCBIs _get_genomes() be run in the background when first called, and only be waited upon when using ncbi features?

I checked, and it could perhaps be done by
* threading
* asynchio

PS I was afraid this would happen for all providers, but it seems the others are really nice and quick!

simonvh commented:
>I don't really like NCBI caching to be the first thing a user sees. If it could be done in the background that would be awesome.

However, any time you initialize a `NCBIProvider`, you probably want to search for genome information, so you need this. Not doing it in `__init__()` means you just defer it to a later point.
I do see the point though, and I also dislike the fact that it takes so long. It is caused by the big file of assembly information. I think the caching maybe not the best thing to do here. We might be able to circumvent this by saving the information in some kind of fast binary format.

>Duplicate output for genomepy search fr3 -p ucsc
I also noticed this for another genome. Strange. Will check.

> Accession IDs are ~15 characters long, while NAs are 2. This causes formatting issues for the search function (ex: genomepy search homo -p ucsc). If the field could always be stretched to a minimum of, say, 20 characters, the issue is circumvented.
Yup. You're right. Some string formatting with padding is necessary here.

simonvh commented:
I've checked different ways of loading/storing data but they're all relatively slow (a few seconds). The caching speed is actually quite good. The main problem (I think?) was in the `list_install_options()` call, which make everything slow for the `genomepy` command. 5512adc is a possible solution, where I made `list_install_options()` a class methods. This means the providers are not instantiated, and don't load the genomes. Let me know what you think.

siebrenf commented:
Formatting isn't consistent yet (try `genomepy search homo -p [provider]` to see the differences). NCBI creates some crazy wide columns.

Not sure if it was just a fluke, but the first minute the UCSC REST API gave a 404.

Not loading the whole NCBI genome when you type `genomepy --help` on first use makes me happy! Its much nicer for the user that the loading of this data only starts when you specifically ask for it :)


New issue: 
now that `search` provides a wealth of nice information, it may not be super obvious which field to use in `genomepy install`. I think this could best be explained in the `genomepy install --help` section(?)

simonvh commented:
>Formatting isn't consistent yet (try genomepy search homo -p [provider] to see the differences). 
> NCBI creates some crazy wide columns.

It is consistent.. always 4 spaces wider per column than the longest entry. However, NCBI indeed has some long, long names. What to about it?

>Not sure if it was just a fluke, but the first minute the UCSC REST API gave a 404.

Seems to have occurred on github too. Looks fine now.

>Not loading the whole NCBI genome when you type genomepy --help on first use makes me 
> happy! Its much nicer for the user that the loading of this data only starts when you specifically ask for it :)

Yay!

> New issue:
> now that search provides a wealth of nice information, it may not be super obvious which field to > use in genomepy install. I think this could best be explained in the genomepy install --help section(?)

Have a look now. I added a header, moved the name to the first columns (which is consistent with how you have to call `genomepy install` later, and added a hint. In case the output is redirected to a file or piped, the hint is not shown.

siebrenf commented:
It looks real nice now! 

With genome names like `GIAB Ashkenazim Mother HG004/NA24143/hu8E87A9 PacBio Assembly with PBcR`  it does indeed get pretty hard to make it nicely formatted...

The last column `other_info` could be left as-is with
```
        sizes = [max(len(row[i]) + 4 for row in data) for i in range(len(data[0])-1)]
        sizes.append(0)
```
This would only make some assemblies take up two lines, but im not sure how much cleaner that is when a row can take up one or two lines.

Otherwise its good to go!

simonvh commented:
> The last column other_info could be left as-is with

Good point. I actually wonder if it needs to be in at all. It contains completely different information for different providers. It might be helpful, but I don't know...

siebrenf commented on genomepy/cli.py:
```diff
@@ -87,8 +112,7 @@ def get_install_options():
     for name in genomepy.provider.ProviderBase.list_providers():
-        p = genomepy.provider.ProviderBase.create(name)
-        p_dict = p.list_install_options()
+        p_dict = genomepy.provider.ProviderBase.list_install_options(name)
```
@simonvh. I'm suddenly getting errors in the annotations branch from this change... No idea why that's only showing up now, but its fixed by undoing the change (actually it's still a one liner: `p_dict = genomepy.provider.ProviderBase.create(name).list_install_options()`). I'm not missing anything right?

PS can this branch be removed now it's been merged?
